server/routerlicious/packages/services-shared/src/redisSocketIoAdapter.ts

```diff
@@ -0,0 +1,506 @@
+    private readonly encoder: any;
+
+    private readonly pendingHealthChecks: Map<string, () => void> = new Map();
+    private readonly roomHealthCheckTimeoutIds: Map<string, NodeJS.Timeout> = new Map();
+
+    /**
+     * Set the Redis connections to use
+     */
+    public static async setup(options: ISocketIoRedisOptions) {
+        this.options = options;
+    }
+
+    constructor(public readonly nsp: socketio.Namespace) {
+        super();
+
+        this.encoder = (nsp.server as any).encoder;
+
+        // todo: better id here?
+        this.uid = uuid.v4().substring(0, 6);
+
+        this.channel = `socket.io#${nsp.name}#`;
+
+        if (this.isDefaultNamespace) {
+            // the default namespace
+            // don't setup stuff for the default namespace. we only use /fluid. this will save memory
+            // related to https://github.com/socketio/socket.io/issues/3089
+            return;
+        }
+    }
+
+    /**
+     * Check if this instance is connected to the default socket io namespace
+     */
+    public get isDefaultNamespace() {
+        return this.nsp.name === "/";
+    }
+
+    /**
+     * Returns the number of unique rooms (not including built in user rooms)
+     */
+    public get uniqueRoomCount() {
+        return this._uniqueRoomCount;
+    }
+
+    /**
+     * Add a socket to a room
+     */
+    public async add(socketId: string, roomId: string, callback?: ((err?: any) => void) | undefined): Promise<void> {
+        return this.addAll(socketId, [roomId], callback);
+    }
+
+    /**
+     * Add a socket to a list of rooms
+     */
+    public async addAll(
+        socketId: string,
+        roomIds: string[],
+        callback?: ((err?: any) => void) | undefined): Promise<void> {
+        if (!this.isDefaultNamespace) {
+            const newRooms: string[] = [];
+
+            for (const roomId of roomIds) {
+                let socketRooms = this.socketIdToRooms.get(socketId);
+                if (!socketRooms) {
+                    socketRooms = new Set();
+                    this.socketIdToRooms.set(socketId, socketRooms);
+                }
+
+                socketRooms.add(roomId);
+
+                let roomSocketIds = this.roomToSocketIds.get(roomId);
+                if (!roomSocketIds) {
+                    roomSocketIds = new Set();
+                    this.roomToSocketIds.set(roomId, roomSocketIds);
+
+                    // don't count the built in user rooms
+                    if (!roomId.startsWith("/")) {
+                        this._uniqueRoomCount++;
+
+                        newRooms.push(roomId);
+                    }
+                }
+
+                roomSocketIds.add(socketId);
+            }
+
+            if (newRooms.length > 0) {
+                // subscribe to the new rooms
+                await this.subscribeToRooms(newRooms);
+            }
+        }
+
+        if (callback) {
+            process.nextTick(callback);
```

Comment:
Why process.nextTick here?

Reply:
This is how the default socket.io-adapter does it so I stayed with it. I have seen bugs in socket.io in the past related to doing work within the same tick as a connection / disconnect event so perhaps this fixes some obscure issue